### ISSUE TITLE
Worker memory leak associated with user-defined transform output function

### ISSUE THREAD

**Issue by paulwanggz** [pipelinedb/pipelinedb#1947]
At present, our pipelinedb database will crash every 8 days.
From day1 to day8, the memory of worker's process will continue to grow.
The following is the memory state before it is in crash:
```
  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND                                                                                                                                                                                    
23066 postgres  25   5 5639884 4.006g 864788 S   1.3 34.6  41:56.16 pipelinedb: bgworker: worker0 [pipeline]                                                                                                                                                   
23064 postgres  25   5 5571340 3.989g 864520 S   0.0 34.5  41:27.72 pipelinedb: bgworker: worker2 [pipeline]                                                                                                                                                   
23065 postgres  25   5 5576028 3.971g 864884 S   0.7 34.3  41:22.79 pipelinedb: bgworker: worker1 [pipeline]  
```
In the log of pipelinedb, you can see the error message as the following:
```
ERROR:  relcache reference ct_rtd_his_narrow_func_fs is not owned by resource owner TopTransaction
STATEMENT:  ct_rtd_his_narrow_func_fs
ERROR:  relcache reference ct_rtd_his_narrow_func_fs is not owned by resource owner TopTransaction
STATEMENT:  ct_rtd_his_narrow_func_fs
ERROR:  relcache reference ct_rtd_his_narrow_func_fs is not owned by resource owner TopTransaction
STATEMENT:  ct_rtd_his_narrow_func_fs
LOG:  worker process: worker0 [pipeline] (PID 23066) was terminated by signal 9: Killed
DETAIL:  Failed process was running: worker0 [pipeline]
LOG:  terminating any other active server processes
WARNING:  terminating connection because of crash of another server process
DETAIL:  The postmaster has commanded this server process to roll back the current transaction and exit, because another server process exited abnormally and possibly corrupted shared memory.
HINT:  In a moment you should be able to reconnect to the database and repeat your command.
WARNING:  terminating connection because of crash of another server process
DETAIL:  The postmaster has commanded this server process to roll back the current transaction and exit, because another server process exited abnormally and possibly corrupted shared memory.
HINT:  In a moment you should be able to reconnect to the database and repeat your command.
WARNING:  terminating connection because of crash of another server process
DETAIL:  The postmaster has commanded this server process to roll back the current transaction and exit, because another server process exited abnormally and possibly corrupted shared memory.
HINT:  In a moment you should be able to reconnect to the database and repeat your command.
WARNING:  terminating connection because of crash of another server process
DETAIL:  The postmaster has commanded this server process to roll back the current transaction and exit, because another server process exited abnormally and possibly corrupted shared memory.
HINT:  In a moment you should be able to reconnect to the database and repeat your command.
WARNING:  terminating connection because of crash of another server process
```
Please note 3 points:
1.In our system, the longest slide Windows (sw) is 15 minutes, and most of the continunue views receive data from the stream according to certain conditions;
2.We use a continunue transfomer to invoke a function(trigger) to perform some actions, such as judging the state of the device or writing data to different tables. There are three similar transformers in my pipelinedb.
3.Please note that the ERROR (relcache reference ct_rtd_his_upper_func_fs is not owned by resource owner TopTransaction) occurs when the function/trigger is called by the continunue transfomer, but this ERROR does not mean the system will crash.
I've been asking about this question at Google group and stack overflow, but there's no feedback.
I would be happy to provide more detailed information.
Hope you can give me some advice to reduce the probability of pipelinedb crash.
Any suggestion will be appreciated:)

**Comment by paulwanggz** [comment]
Supplementary information:
1. my version of Pipelinedb:
pipeline=# SELECT pipeline_version();    
                                                                       pipeline_version                                                                        
---------------------------------------------------------------------------------------------------------------------------------------------------------------
 PipelineDB 0.9.9 at revision 6369f5b0615f3643d1e5613fe0433532ce9092f7 on x86_64-pc-linux-gnu, compiled by gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-11), 64-bit
(1 row)
2.My system environment: centos 7.3/12g memory/4Core

**Comment by derekjn** [comment]
@paulwanggz thank you for this writeup! We suspect there is a bug with user-defined functions used with transforms, which is causing both the errors you're seeing as well as a memory leak. We'll report back when we know more.

**Comment by derekjn** [comment]
From the original Google Groups thread:

```sql
CREATE CONTINUOUS TRANSFORM chr.ct_mn_ver_change_fs AS
  SELECT (new).mn_code,(new).arvl_time,
         (old).dev as old_dev,
         (new).dev  as new_dev,
         (old).usg as old_usg,
         (new).usg  as new_usg,
         (old).sver as old_sver,
         (new).sver as new_sver,
         (old).hver as old_hver,
         (new).hver as new_hver
  FROM output_of('chr.cv_mn_ver_cur_fs')
  WHERE COALESCE((new).dev,'') <> COALESCE((old).dev,'')
     OR COALESCE((new).usg,'') <> COALESCE((old).usg,'')
     OR COALESCE((new).sver,'') <> COALESCE((old).sver,'')
     OR COALESCE((new).hver,'') <> COALESCE((old).hver,'')
THEN EXECUTE PROCEDURE chr.f_modify_mn_ver_fs()
```

```sql
CREATE OR REPLACE FUNCTION chr.f_modify_mn_ver_fs()
  RETURNS trigger AS
  $$
  DECLARE
      v_rowcount  int;
  BEGIN
  	 RAISE NOTICE '=== f_modify_mn_ver_fs ===';
  	 RAISE NOTICE 'iphone';
     RAISE NOTICE '%,%,%,%,%,%,%,%,%', NEW.mn_code, NEW.arvl_time, NEW.old_dev, NEW.new_dev, NEW.old_sver, NEW.new_sver, NEW.old_hver, NEW.new_hver, clock_timestamp();

     IF COALESCE(NEW.new_dev,'') <> COALESCE(NEW.old_dev,'') THEN
        SELECT count(1) INTO v_rowcount FROM chr.t_mn_ver_fs WHERE mn_code=NEW.mn_code AND ver_type='dev' AND is_current='Y';
        IF v_rowcount > 0 THEN
            UPDATE chr.t_mn_ver
                  SET end_time = NEW.arvl_time,
                      is_current = 'N',
                      update_time = clock_timestamp()
                  WHERE mn_code = NEW.mn_code
                  AND ver_type='dev'
                  AND is_current = 'Y';
        END IF;

        INSERT INTO chr.t_mn_ver_fs (mn_code, ver_type, ver, begin_time, end_time, is_current, insert_time)
                    VALUES (NEW.mn_code, 'dev', NEW.new_sver, NEW.arvl_time, NULL, 'Y', clock_timestamp());
      END IF;

      IF COALESCE(NEW.new_usg,'') <> COALESCE(NEW.old_usg,'') THEN
        SELECT count(1) INTO v_rowcount FROM chr.t_mn_ver_fs WHERE mn_code=NEW.mn_code AND ver_type='usg' AND is_current='Y';
        IF v_rowcount > 0 THEN
            UPDATE chr.t_mn_ver_fs
                  SET end_time = NEW.arvl_time,
                      is_current = 'N',
                      update_time = clock_timestamp()
                  WHERE mn_code = NEW.mn_code
                  AND ver_type='usg'
                  AND is_current = 'Y';
        END IF;

        INSERT INTO chr.t_mn_ver_fs (mn_code, ver_type, ver, begin_time, end_time, is_current, insert_time)
                    VALUES (NEW.mn_code, 'usg', NEW.new_usg, NEW.arvl_time, NULL, 'Y', clock_timestamp());
      END IF;

      IF COALESCE(NEW.new_sver,'') <> COALESCE(NEW.old_sver,'') THEN
          SELECT count(1) INTO v_rowcount FROM chr.t_mn_ver_fs WHERE mn_code=NEW.mn_code AND ver_type='sver' AND is_current='Y';
          IF v_rowcount > 0 THEN
            UPDATE chr.t_mn_ver_fs
                  SET end_time = NEW.arvl_time,
                      is_current = 'N',
                      update_time = clock_timestamp()
                  WHERE mn_code = NEW.mn_code
                  AND ver_type='sver'
                  AND is_current = 'Y';
          END IF;

          INSERT INTO chr.t_mn_ver_fs (mn_code, ver_type, ver, begin_time, end_time, is_current, insert_time)
                    VALUES (NEW.mn_code, 'sver', NEW.new_sver, NEW.arvl_time, NULL, 'Y', clock_timestamp());
      END IF;
      IF COALESCE(NEW.new_hver,'') <> COALESCE(NEW.old_hver,'') THEN
          SELECT count(1) INTO v_rowcount FROM chr.t_mn_ver_fs WHERE mn_code=NEW.mn_code AND ver_type='hver' AND is_current='Y';
          IF v_rowcount > 0 THEN
            UPDATE chr.t_mn_ver
                  SET end_time = NEW.arvl_time,
                      is_current = 'N',
                      update_time = clock_timestamp()
                  WHERE mn_code = NEW.mn_code
                  AND ver_type='hver'
                  AND is_current = 'Y';
          END IF;

          INSERT INTO chr.t_mn_ver_fs (mn_code, ver_type, ver, begin_time, end_time, is_current, insert_time)
                    VALUES (NEW.mn_code, 'hver', NEW.new_hver, NEW.arvl_time, NULL, 'Y', clock_timestamp());

     END IF;

  RETURN NEW;
  END;
  $$
  LANGUAGE plpgsql;
```

**Comment by derekjn** [comment]
@paulwanggz do you mind posting here the SQL used to create the remaining objects? Looks like these are missing:

* `cv_mn_ver_cur_fs`
* `t_mn_ver_fs`

**Comment by paulwanggz** [comment]
@derekjn 

DROP TABLE chr.t_mn_status_fs;
CREATE TABLE chr.t_mn_status_fs(
    mn_code         varchar(100),
    begin_time      timestamp(0) without time zone,
    end_time        timestamp(0) without time zone,
    status          char(2),
    is_current      char(2),
    is_delete       char(1) default 'N',
    insert_time     timestamp,
    update_time     timestamp
);

DROP CONTINUOUS VIEW chr.cv_mn_ver_cur_fs;
CREATE CONTINUOUS VIEW chr.cv_mn_ver_cur_fs AS
   SELECT (data_value->>'mn_code')::varchar(32) AS mn_code,
          max(arrival_timestamp) AS arvl_time,
   				keyed_max(arrival_timestamp,(data_value->'tag'->>'DEV')::varchar(32)) as dev,
   				keyed_max(arrival_timestamp,(data_value->'tag'->>'USAGE')::varchar(32)) as usg,
   				keyed_max(arrival_timestamp,(data_value->'metric'->>'SVER')::varchar(32)) as sver,
   				keyed_max(arrival_timestamp,(data_value->'metric'->>'HVER')::varchar(32)) as hver
     FROM chr.strm_mgt_data_packet_fs
   WHERE (data_value->'tag'->>'CN')::varchar(8)='8011'
   GROUP BY (data_value->>'mn_code')::varchar(32)
;

**Comment by derekjn** [comment]
@paulwanggz thank you very much!

**Comment by paulwanggz** [comment]
@derekjn Maybe you still need the data example.
```
{"cn": "8011", "tag": {"CN": "8011", "MN": "88882205280049", "PW": "123456", "QN": "20180727220658", "ST": "22", "DEV": "B280V3", "Flag": "1", "USAGE": "001"}, "metric": {"Rtd": {"BV": "0", "LG": "0", "LT": "0", "LGD": "0", "L
TD": "0", "SPD": "0", "KB03": "0", "KP01": "0", "KS01": "0", "KW01": "0", "KPM10": "0", "KPM25": "0"}, "Flag": {"BV": "N", "LG": "N", "LT": "N", "LGD": "N", "LTD": "N", "SPD": "N", "KB03": "N", "KP01": "N", "KS01": "N", "KW01": "N", "KPM10": "N", "KPM25": 
"N"}, "HVER": "V3.0.0", "SVER": "V3.0.5R", "RngMax": {"KB03": "9999.99", "KP01": "9999.99", "KS01": "9999.99", "KW01": "9999.99", "KPM10": "9999.99", "KPM25": "9999.99"}, "RngMin": {"KB03": "0.0", "KP01": "0.0", "KS01": "0.0", "KW01": "0.0", "KPM10": "0.0"
, "KPM25": "0.0"}, "LoginTime": "20180727220658"}, "mn_code": "88882205280049", "data_time": "2018-07-27 22:06:58.0", "recv_time": "2018-07-27 22:07:13.307"}
```

**Comment by derekjn** [comment]
@paulwanggz we're just about done looking into this, can you give us the definition of `chr.t_mn_ver_fs`?

**Comment by derekjn** [comment]
Consolidated DDL:

```sql
DROP SCHEMA chr CASCADE;
CREATE SCHEMA chr;

CREATE FOREIGN TABLE chr.strm_mgt_data_packet_fs (data_value json) SERVER pipelinedb;

CREATE OR REPLACE FUNCTION chr.f_modify_mn_ver_fs()
  RETURNS trigger AS
  $$
  DECLARE
      v_rowcount  int;
  BEGIN
  	 RAISE NOTICE '=== f_modify_mn_ver_fs ===';
  	 RAISE NOTICE 'iphone';
     RAISE NOTICE '%,%,%,%,%,%,%,%,%', NEW.mn_code, NEW.arvl_time, NEW.old_dev, NEW.new_dev, NEW.old_sver, NEW.new_sver, NEW.old_hver, NEW.new_hver, clock_timestamp();

     IF COALESCE(NEW.new_dev,'') <> COALESCE(NEW.old_dev,'') THEN
        SELECT count(1) INTO v_rowcount FROM chr.t_mn_ver_fs WHERE mn_code=NEW.mn_code AND ver_type='dev' AND is_current='Y';
        IF v_rowcount > 0 THEN
            UPDATE chr.t_mn_ver
                  SET end_time = NEW.arvl_time,
                      is_current = 'N',
                      update_time = clock_timestamp()
                  WHERE mn_code = NEW.mn_code
                  AND ver_type='dev'
                  AND is_current = 'Y';
        END IF;

        INSERT INTO chr.t_mn_ver_fs (mn_code, ver_type, ver, begin_time, end_time, is_current, insert_time)
                    VALUES (NEW.mn_code, 'dev', NEW.new_sver, NEW.arvl_time, NULL, 'Y', clock_timestamp());
      END IF;

      IF COALESCE(NEW.new_usg,'') <> COALESCE(NEW.old_usg,'') THEN
        SELECT count(1) INTO v_rowcount FROM chr.t_mn_ver_fs WHERE mn_code=NEW.mn_code AND ver_type='usg' AND is_current='Y';
        IF v_rowcount > 0 THEN
            UPDATE chr.t_mn_ver_fs
                  SET end_time = NEW.arvl_time,
                      is_current = 'N',
                      update_time = clock_timestamp()
                  WHERE mn_code = NEW.mn_code
                  AND ver_type='usg'
                  AND is_current = 'Y';
        END IF;

        INSERT INTO chr.t_mn_ver_fs (mn_code, ver_type, ver, begin_time, end_time, is_current, insert_time)
                    VALUES (NEW.mn_code, 'usg', NEW.new_usg, NEW.arvl_time, NULL, 'Y', clock_timestamp());
      END IF;

      IF COALESCE(NEW.new_sver,'') <> COALESCE(NEW.old_sver,'') THEN
          SELECT count(1) INTO v_rowcount FROM chr.t_mn_ver_fs WHERE mn_code=NEW.mn_code AND ver_type='sver' AND is_current='Y';
          IF v_rowcount > 0 THEN
            UPDATE chr.t_mn_ver_fs
                  SET end_time = NEW.arvl_time,
                      is_current = 'N',
                      update_time = clock_timestamp()
                  WHERE mn_code = NEW.mn_code
                  AND ver_type='sver'
                  AND is_current = 'Y';
          END IF;

          INSERT INTO chr.t_mn_ver_fs (mn_code, ver_type, ver, begin_time, end_time, is_current, insert_time)
                    VALUES (NEW.mn_code, 'sver', NEW.new_sver, NEW.arvl_time, NULL, 'Y', clock_timestamp());
      END IF;
      IF COALESCE(NEW.new_hver,'') <> COALESCE(NEW.old_hver,'') THEN
          SELECT count(1) INTO v_rowcount FROM chr.t_mn_ver_fs WHERE mn_code=NEW.mn_code AND ver_type='hver' AND is_current='Y';
          IF v_rowcount > 0 THEN
            UPDATE chr.t_mn_ver
                  SET end_time = NEW.arvl_time,
                      is_current = 'N',
                      update_time = clock_timestamp()
                  WHERE mn_code = NEW.mn_code
                  AND ver_type='hver'
                  AND is_current = 'Y';
          END IF;

          INSERT INTO chr.t_mn_ver_fs (mn_code, ver_type, ver, begin_time, end_time, is_current, insert_time)
                    VALUES (NEW.mn_code, 'hver', NEW.new_hver, NEW.arvl_time, NULL, 'Y', clock_timestamp());

     END IF;

  RETURN NEW;
  END;
  $$
  LANGUAGE plpgsql;

CREATE VIEW chr.cv_mn_ver_cur_fs AS
   SELECT (data_value->>'mn_code')::varchar(32) AS mn_code,
          max(arrival_timestamp) AS arvl_time,
   	  keyed_max(arrival_timestamp,(data_value->'tag'->>'DEV')::varchar(32)) as dev,
   	  keyed_max(arrival_timestamp,(data_value->'tag'->>'USAGE')::varchar(32)) as usg,
   	  keyed_max(arrival_timestamp,(data_value->'metric'->>'SVER')::varchar(32)) as sver,
   	  keyed_max(arrival_timestamp,(data_value->'metric'->>'HVER')::varchar(32)) as hver
     FROM chr.strm_mgt_data_packet_fs
   WHERE (data_value->'tag'->>'CN')::varchar(8)='8011'
   GROUP BY (data_value->>'mn_code')::varchar(32);

CREATE VIEW chr.ct_mn_ver_change_fs WITH (action=transform, outputfunc=chr.f_modify_mn_ver_fs) AS
  SELECT (new).mn_code,(new).arvl_time,
         (old).dev as old_dev,
         (new).dev  as new_dev,
         (old).usg as old_usg,
         (new).usg  as new_usg,
         (old).sver as old_sver,
         (new).sver as new_sver,
         (old).hver as old_hver,
         (new).hver as new_hver
  FROM output_of('chr.cv_mn_ver_cur_fs')
  WHERE COALESCE((new).dev,'') <> COALESCE((old).dev,'')
     OR COALESCE((new).usg,'') <> COALESCE((old).usg,'')
     OR COALESCE((new).sver,'') <> COALESCE((old).sver,'')
     OR COALESCE((new).hver,'') <> COALESCE((old).hver,'');
```

**Comment by derekjn** [comment]
@paulwanggz this is fixed in the upcoming `1.0.0` release. Thanks for providing all of this information!
